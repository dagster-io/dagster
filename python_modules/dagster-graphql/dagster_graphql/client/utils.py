from enum import Enum
from typing import Any, NamedTuple, Optional


class DagsterGraphQLClientError(Exception):
    def __init__(self, *args, body=None):
        super().__init__(*args)
        self.body = body


class ReloadRepositoryLocationStatus(Enum):
    """This enum describes the status of a GraphQL mutation to reload a Dagster repository location.

    Args:
        Enum (str): can be either `ReloadRepositoryLocationStatus.SUCCESS`
            or `ReloadRepositoryLocationStatus.FAILURE`.
    """

    SUCCESS = "SUCCESS"
    FAILURE = "FAILURE"


class ShutdownRepositoryLocationStatus(Enum):
    SUCCESS = "SUCCESS"
    FAILURE = "FAILURE"


class ReloadRepositoryLocationInfo(NamedTuple):
    """This class gives information about the result of reloading
    a Dagster repository location with a GraphQL mutation.

    Args:
        status (ReloadRepositoryLocationStatus): The status of the reload repository location mutation
        failure_type: (Optional[str], optional): the failure type if `status == ReloadRepositoryLocationStatus.FAILURE`.
          Can be one of `ReloadNotSupported`, `RepositoryLocationNotFound`, or `RepositoryLocationLoadFailure`. Defaults to None.
        message (Optional[str], optional): the failure message/reason if
          `status == ReloadRepositoryLocationStatus.FAILURE`. Defaults to None.
    """

    status: ReloadRepositoryLocationStatus
    failure_type: Optional[str] = None
    message: Optional[str] = None


class ShutdownRepositoryLocationInfo(NamedTuple):
    """This class gives information about the result of shutting down the server for
    a Dagster repository location using a GraphQL mutation.

    Args:
        status (ShutdownRepositoryLocationStatus) Whether the shutdown succeeded or failed.
        message (Optional[str], optional): the failure message/reason if
          `status == ShutdownRepositoryLocationStatus.FAILURE`. Defaults to None.
    """

    status: ShutdownRepositoryLocationStatus
    message: Optional[str] = None


class JobInfo(NamedTuple):
    repository_location_name: str
    repository_name: str
    job_name: str

    @staticmethod
    def from_node(node: dict[str, Any]) -> list["JobInfo"]:
        repo_name = node["name"]
        repo_location_name = node["location"]["name"]
        return [
            JobInfo(
                repository_location_name=repo_location_name,
                repository_name=repo_name,
                job_name=job["name"],
            )
            for job in node["pipelines"]
        ]


class InvalidOutputErrorInfo(NamedTuple):
    """This class gives information about an InvalidOutputError from submitting a pipeline for execution
    from GraphQL.

    Args:
        step_key (str): key of the step that failed
        invalid_output_name (str): the name of the invalid output from the given step
    """

    step_key: str
    invalid_output_name: str
