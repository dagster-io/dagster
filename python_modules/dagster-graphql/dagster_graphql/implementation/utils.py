import functools
import sys
from collections.abc import Iterable, Iterator, Mapping, Sequence
from contextlib import contextmanager
from contextvars import ContextVar
from datetime import datetime
from types import TracebackType
from typing import TYPE_CHECKING, Any, Callable, NamedTuple, Optional, TypeVar, Union, cast

import dagster._check as check
from dagster._core.definitions.asset_check_spec import AssetCheckKey
from dagster._core.definitions.events import AssetKey
from dagster._core.definitions.partition import PartitionsDefinition
from dagster._core.definitions.remote_asset_graph import RemoteWorkspaceAssetGraph
from dagster._core.definitions.selector import GraphSelector, JobSubsetSelector
from dagster._core.execution.backfill import PartitionBackfill
from dagster._core.workspace.context import BaseWorkspaceRequestContext
from dagster._utils.caching_instance_queryer import CachingInstanceQueryer
from dagster._utils.error import serializable_error_info_from_exc_info
from typing_extensions import ParamSpec, TypeAlias

if TYPE_CHECKING:
    from dagster_graphql.schema.errors import GrapheneError, GraphenePythonError
    from dagster_graphql.schema.util import ResolveInfo

P = ParamSpec("P")
T = TypeVar("T")

GrapheneResolverFn: TypeAlias = Callable[..., object]
T_Callable = TypeVar("T_Callable", bound=Callable)


def assert_permission_for_location(
    graphene_info: "ResolveInfo", permission: str, location_name: str
) -> None:
    from dagster_graphql.schema.errors import GrapheneUnauthorizedError

    context = cast(BaseWorkspaceRequestContext, graphene_info.context)
    if not context.has_permission_for_location(permission, location_name):
        raise UserFacingGraphQLError(GrapheneUnauthorizedError())


def require_permission_check(
    permission: str,
) -> Callable[[GrapheneResolverFn], GrapheneResolverFn]:
    def decorator(fn: GrapheneResolverFn) -> GrapheneResolverFn:
        @functools.wraps(fn)
        def _fn(self, graphene_info, *args: P.args, **kwargs: P.kwargs):
            result = fn(self, graphene_info, *args, **kwargs)

            if not graphene_info.context.was_permission_checked(permission):
                raise Exception(f"Permission {permission} was never checked during the request")

            return result

        return _fn

    return decorator


def check_permission(
    permission: str,
) -> Callable[[GrapheneResolverFn], GrapheneResolverFn]:
    def decorator(fn: GrapheneResolverFn) -> GrapheneResolverFn:
        @functools.wraps(fn)
        def _fn(self, graphene_info, *args: P.args, **kwargs: P.kwargs):
            assert_permission(graphene_info, permission)

            return fn(self, graphene_info, *args, **kwargs)

        return _fn

    return decorator


def assert_permission(graphene_info: "ResolveInfo", permission: str) -> None:
    from dagster_graphql.schema.errors import GrapheneUnauthorizedError

    context = cast(BaseWorkspaceRequestContext, graphene_info.context)
    if not context.has_permission(permission):
        raise UserFacingGraphQLError(GrapheneUnauthorizedError())


def has_permission_for_asset_graph(
    graphene_info: "ResolveInfo",
    asset_graph: RemoteWorkspaceAssetGraph,
    asset_selection: Optional[Sequence[AssetKey]],
    permission: str,
) -> bool:
    asset_keys = set(asset_selection or [])
    context = cast(BaseWorkspaceRequestContext, graphene_info.context)

    if asset_keys:
        location_names = set()
        for key in asset_keys:
            if not asset_graph.has(key):
                # If any of the asset keys don't map to a location (e.g. because they are no longer in the
                # graph) need deployment-wide permissions - no valid code location to check
                return context.has_permission(permission)
            node = asset_graph.get(key)
            location_names.add(
                node.resolve_to_singular_repo_scoped_node().repository_handle.location_name
            )
    else:
        location_names = set(
            handle.location_name for handle in asset_graph.repository_handles_by_key.values()
        )

    if not location_names:
        return context.has_permission(permission)
    else:
        return all(
            context.has_permission_for_location(permission, location_name)
            for location_name in location_names
        )


def assert_permission_for_asset_graph(
    graphene_info: "ResolveInfo",
    asset_graph: RemoteWorkspaceAssetGraph,
    asset_selection: Optional[Sequence[AssetKey]],
    permission: str,
) -> None:
    from dagster_graphql.schema.errors import GrapheneUnauthorizedError

    if not has_permission_for_asset_graph(graphene_info, asset_graph, asset_selection, permission):
        raise UserFacingGraphQLError(GrapheneUnauthorizedError())


def assert_valid_job_partition_backfill(
    graphene_info: "ResolveInfo",
    backfill: PartitionBackfill,
    partitions_def: PartitionsDefinition,
    dynamic_partitions_store: CachingInstanceQueryer,
    backfill_datetime: datetime,
) -> None:
    from dagster_graphql.schema.errors import GraphenePartitionKeysNotFoundError

    partition_names = backfill.get_partition_names(graphene_info.context)

    if not partition_names:
        return

    invalid_keys = set(partition_names) - set(
        partitions_def.get_partition_keys(backfill_datetime, dynamic_partitions_store)
    )

    if invalid_keys:
        raise UserFacingGraphQLError(GraphenePartitionKeysNotFoundError(invalid_keys))


def assert_valid_asset_partition_backfill(
    graphene_info: "ResolveInfo",
    backfill: PartitionBackfill,
    dynamic_partitions_store: CachingInstanceQueryer,
    backfill_datetime: datetime,
) -> None:
    from dagster_graphql.schema.errors import GraphenePartitionKeysNotFoundError

    asset_graph = graphene_info.context.asset_graph
    asset_backfill_data = backfill.asset_backfill_data

    if not asset_backfill_data:
        return

    partition_subset_by_asset_key = (
        asset_backfill_data.target_subset.partitions_subsets_by_asset_key
    )

    for asset_key, partition_subset in partition_subset_by_asset_key.items():
        partitions_def = asset_graph.get(asset_key).partitions_def

        if not partitions_def:
            continue

        invalid_keys = set(partition_subset.get_partition_keys()) - set(
            partitions_def.get_partition_keys(backfill_datetime, dynamic_partitions_store)
        )

        if invalid_keys:
            raise UserFacingGraphQLError(GraphenePartitionKeysNotFoundError(invalid_keys))


def _noop(_) -> None:
    pass


class ErrorCapture:
    @staticmethod
    def default_on_exception(
        exc_info: tuple[type[BaseException], BaseException, TracebackType],
    ) -> "GraphenePythonError":
        from dagster_graphql.schema.errors import GraphenePythonError

        # Transform exception in to PythonError to present to user
        return GraphenePythonError(serializable_error_info_from_exc_info(exc_info))

    # global behavior for how to handle unexpected exceptions
    on_exception = default_on_exception

    # context var for observing unexpected exceptions
    observer: ContextVar[Callable[[Exception], None]] = ContextVar(
        "error_capture_observer", default=_noop
    )

    @staticmethod
    @contextmanager
    def watch(fn: Callable[[Exception], None]) -> Iterator[None]:
        token = ErrorCapture.observer.set(fn)
        try:
            yield
        finally:
            ErrorCapture.observer.reset(token)


def capture_error(
    fn: Callable[P, T],
) -> Callable[P, Union[T, "GrapheneError", "GraphenePythonError"]]:
    @functools.wraps(fn)
    def _fn(*args: P.args, **kwargs: P.kwargs) -> T:
        try:
            return fn(*args, **kwargs)
        except UserFacingGraphQLError as de_exception:
            return de_exception.error
        except Exception as exc:
            ErrorCapture.observer.get()(exc)
            return ErrorCapture.on_exception(sys.exc_info())  # type: ignore

    return _fn


class UserFacingGraphQLError(Exception):
    # The `error` arg here should be a Graphene type implementing the interface `GrapheneError`, but
    # this is not trackable by the Python type system.
    def __init__(self, error: Any):
        self.error = error
        message = "[{cls}] {message}".format(
            cls=error.__class__.__name__,
            message=error.message if hasattr(error, "message") else None,
        )
        super().__init__(message)


def pipeline_selector_from_graphql(data: Mapping[str, Any]) -> JobSubsetSelector:
    asset_selection = cast(Optional[Iterable[dict[str, list[str]]]], data.get("assetSelection"))
    asset_check_selection = cast(
        Optional[Iterable[dict[str, Any]]], data.get("assetCheckSelection")
    )
    return JobSubsetSelector(
        location_name=data["repositoryLocationName"],
        repository_name=data["repositoryName"],
        job_name=data.get("pipelineName") or data.get("jobName"),  # type: ignore
        op_selection=data.get("solidSelection"),
        asset_selection=(
            [AssetKey.from_graphql_input(asset_key) for asset_key in asset_selection]
            if asset_selection
            else None
        ),
        asset_check_selection=(
            [AssetCheckKey.from_graphql_input(asset_check) for asset_check in asset_check_selection]
            if asset_check_selection is not None
            else None
        ),
    )


def graph_selector_from_graphql(data: Mapping[str, Any]) -> GraphSelector:
    return GraphSelector(
        location_name=data["repositoryLocationName"],
        repository_name=data["repositoryName"],
        graph_name=data["graphName"],
    )


class ExecutionParams(
    NamedTuple(
        "_ExecutionParams",
        [
            ("selector", JobSubsetSelector),
            ("run_config", Mapping[str, object]),
            ("mode", Optional[str]),
            ("execution_metadata", "ExecutionMetadata"),
            ("step_keys", Optional[Sequence[str]]),
        ],
    )
):
    def __new__(
        cls,
        selector: JobSubsetSelector,
        run_config: Optional[Mapping[str, object]],
        mode: Optional[str],
        execution_metadata: "ExecutionMetadata",
        step_keys: Optional[Sequence[str]],
    ):
        check.opt_list_param(step_keys, "step_keys", of_type=str)

        return super().__new__(
            cls,
            selector=check.inst_param(selector, "selector", JobSubsetSelector),
            run_config=check.opt_mapping_param(run_config, "run_config", key_type=str),
            mode=check.opt_str_param(mode, "mode"),
            execution_metadata=check.inst_param(
                execution_metadata, "execution_metadata", ExecutionMetadata
            ),
            step_keys=step_keys,
        )

    def to_graphql_input(self) -> Mapping[str, Any]:
        return {
            "selector": self.selector.to_graphql_input(),
            "runConfigData": self.run_config,
            "mode": self.mode,
            "executionMetadata": self.execution_metadata.to_graphql_input(),
            "stepKeys": self.step_keys,
        }


class ExecutionMetadata(
    NamedTuple(
        "_ExecutionMetadata",
        [
            ("run_id", Optional[str]),
            ("tags", Mapping[str, str]),
            ("root_run_id", Optional[str]),
            ("parent_run_id", Optional[str]),
        ],
    )
):
    def __new__(
        cls,
        run_id: Optional[str],
        tags: Mapping[str, str],
        root_run_id: Optional[str] = None,
        parent_run_id: Optional[str] = None,
    ):
        return super().__new__(
            cls,
            check.opt_str_param(run_id, "run_id"),
            check.dict_param(tags, "tags", key_type=str, value_type=str),
            check.opt_str_param(root_run_id, "root_run_id"),
            check.opt_str_param(parent_run_id, "parent_run_id"),
        )

    def to_graphql_input(self) -> Mapping[str, Any]:
        return {
            "runId": self.run_id,
            "tags": [{"key": k, "value": v} for k, v in self.tags.items()],
            "rootRunId": self.root_run_id,
            "parentRunId": self.parent_run_id,
        }


def apply_cursor_limit_reverse(
    items: Sequence[str],
    cursor: Optional[str],
    limit: Optional[int],
    reverse: Optional[bool],
) -> Sequence[str]:
    start = 0
    end = len(items)
    index = 0

    if cursor:
        index = next(idx for (idx, item) in enumerate(items) if item == cursor)

        if reverse:
            end = index
        else:
            start = index + 1

    if limit:
        if reverse:
            start = end - limit
        else:
            end = start + limit

    return items[max(start, 0) : end]


BackfillParams: TypeAlias = Mapping[str, Any]
AssetBackfillPreviewParams: TypeAlias = Mapping[str, Any]
