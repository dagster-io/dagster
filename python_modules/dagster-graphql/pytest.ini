[pytest]
markers =
    variant:
    sqlite_instance: Uses a sqlite instance.
    postgres_instance: Uses a postgres instance.
    sync_run_launcher: Uses the synchronous run launcher
    default_run_launcher: User the default run launcher
    queued_run_coordinator: Uses the QueuedRunCoordinator
    read_only: The context is in readonly mode
    non_launchable: The instance does has neither a run launcher nor an execution manager configured
    multi_location: Workspace has multiple repository locations
    managed_grpc_env: Dagster creates a gRPC server to run user code
    deployed_grpc_env: User is running their own gRPC server via `dagster api grpc`
    code_server_cli_grpc_env: User is running their own gRPC server via `dagster code-server start`
    graphql_context_test_suite: Uses graphql_context_test_suite infrastructure.
    asset_aware_instance:
    python_client_test_suite: Tests for the GraphQL Python client

filterwarnings =
    ignore::dagster._utils.warnings.BetaWarning
    ignore::dagster._utils.warnings.PreviewWarning
    ignore::dagster._utils.warnings.SupersessionWarning
