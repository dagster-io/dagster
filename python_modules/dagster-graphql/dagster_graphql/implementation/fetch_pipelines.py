from typing import TYPE_CHECKING, Union

import dagster._check as check
from dagster._core.instance import DagsterInstance
from dagster._core.storage.dagster_run import DagsterRun

from dagster_graphql.implementation.external import get_full_remote_job_or_raise
from dagster_graphql.implementation.utils import JobSubsetSelector, UserFacingGraphQLError
from dagster_graphql.schema.util import ResolveInfo

if TYPE_CHECKING:
    from dagster_graphql.schema.pipelines.pipeline_ref import GrapheneUnknownPipeline
    from dagster_graphql.schema.pipelines.snapshot import GraphenePipelineSnapshot


async def get_job_snapshot_or_error_from_job_selector(
    graphene_info: ResolveInfo, job_selector: JobSubsetSelector
) -> "GraphenePipelineSnapshot":
    from dagster_graphql.schema.pipelines.snapshot import GraphenePipelineSnapshot

    check.inst_param(job_selector, "pipeline_selector", JobSubsetSelector)
    return GraphenePipelineSnapshot(await get_full_remote_job_or_raise(graphene_info, job_selector))


def get_job_snapshot_or_error_from_snapshot_id(
    graphene_info: ResolveInfo, snapshot_id: str
) -> "GraphenePipelineSnapshot":
    check.str_param(snapshot_id, "snapshot_id")

    return _get_job_snapshot_from_instance(graphene_info.context.instance, snapshot_id)


async def get_job_snapshot_or_error_from_snap_or_selector(
    graphene_info: ResolveInfo,
    job_selector: JobSubsetSelector,
    snapshot_id: str,
):
    from dagster_graphql.schema.pipelines.snapshot import GraphenePipelineSnapshot

    if graphene_info.context.instance.has_job_snapshot(snapshot_id):
        job_snapshot = graphene_info.context.instance.get_historical_job(snapshot_id)
        if job_snapshot:
            return GraphenePipelineSnapshot(job_snapshot)

    return await get_job_snapshot_or_error_from_job_selector(graphene_info, job_selector)


# extracted this out to test
def _get_job_snapshot_from_instance(
    instance: DagsterInstance, snapshot_id: str
) -> "GraphenePipelineSnapshot":
    from dagster_graphql.schema.errors import GraphenePipelineSnapshotNotFoundError
    from dagster_graphql.schema.pipelines.snapshot import GraphenePipelineSnapshot

    if not instance.has_job_snapshot(snapshot_id):
        raise UserFacingGraphQLError(GraphenePipelineSnapshotNotFoundError(snapshot_id))

    historical_pipeline = instance.get_historical_job(snapshot_id)

    if not historical_pipeline:
        # Either a temporary error or it has been deleted in the interim
        raise UserFacingGraphQLError(GraphenePipelineSnapshotNotFoundError(snapshot_id))

    return GraphenePipelineSnapshot(historical_pipeline)


def get_job_reference_or_raise(
    graphene_info: ResolveInfo, dagster_run: DagsterRun
) -> Union["GraphenePipelineSnapshot", "GrapheneUnknownPipeline"]:
    """Returns a PipelineReference or raises a UserFacingGraphQLError if a pipeline
    reference cannot be retrieved based on the run, e.g, a UserFacingGraphQLError that wraps an
    InvalidSubsetError.
    """
    from dagster_graphql.schema.pipelines.pipeline_ref import GrapheneUnknownPipeline

    check.inst_param(dagster_run, "pipeline_run", DagsterRun)
    op_selection = (
        list(dagster_run.resolved_op_selection) if dagster_run.resolved_op_selection else None
    )

    if dagster_run.job_snapshot_id is None:
        return GrapheneUnknownPipeline(dagster_run.job_name, op_selection)

    return _get_job_snapshot_from_instance(
        graphene_info.context.instance, dagster_run.job_snapshot_id
    )
