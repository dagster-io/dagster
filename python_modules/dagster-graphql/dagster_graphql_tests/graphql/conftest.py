import tempfile

import pytest
from dagster._core.test_utils import instance_for_test

from dagster_graphql_tests.graphql.repo import define_test_out_of_process_context
from dagster_graphql_tests.graphql.repo_definitions import (
    define_definitions_test_out_of_process_context,
)


@pytest.yield_fixture(scope="function")
def graphql_context():
    with tempfile.TemporaryDirectory() as temp_dir:
        with instance_for_test(
            temp_dir=temp_dir,
            overrides={
                "scheduler": {
                    "module": "dagster.utils.test",
                    "class": "FilesystemTestScheduler",
                    "config": {"base_dir": temp_dir},
                },
                "run_coordinator": {
                    "module": "dagster._core.run_coordinator.immediately_launch_run_coordinator",
                    "class": "ImmediatelyLaunchRunCoordinator",
                },
            },
        ) as instance:
            with define_test_out_of_process_context(instance) as context:
                yield context


@pytest.yield_fixture(scope="function")
def definitions_graphql_context():
    with tempfile.TemporaryDirectory() as temp_dir:
        with instance_for_test(
            temp_dir=temp_dir,
            overrides={
                "scheduler": {
                    "module": "dagster.utils.test",
                    "class": "FilesystemTestScheduler",
                    "config": {"base_dir": temp_dir},
                }
            },
        ) as instance:
            with define_definitions_test_out_of_process_context(instance) as context:
                yield context
