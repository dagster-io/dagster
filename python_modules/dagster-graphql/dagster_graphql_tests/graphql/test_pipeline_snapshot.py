from unittest import mock

import pytest
from dagster._core.workspace.context import WorkspaceRequestContext
from dagster_graphql.implementation.fetch_pipelines import _get_job_snapshot_from_instance
from dagster_graphql.implementation.utils import UserFacingGraphQLError
from dagster_graphql.test.utils import (
    execute_dagster_graphql,
    main_repo_location_name,
    main_repo_name,
)
from dagster_shared.seven import json

from dagster_graphql_tests.graphql.repo import noop_job

SNAPSHOT_OR_ERROR_QUERY_BY_SNAPSHOT_ID = """
query PipelineSnapshotQueryBySnapshotID($snapshotId: String!) {
    pipelineSnapshotOrError(snapshotId: $snapshotId) {
        __typename
        ... on PipelineSnapshot {
            name
            pipelineSnapshotId
            description
            dagsterTypes { key }
            solids { name }
            modes { name }
            solidHandles { handleID }
            tags { key value }
            runTags { key value }
        }
        ... on PipelineSnapshotNotFoundError {
            snapshotId
        }
    }
}
"""

SNAPSHOT_OR_ERROR_QUERY_BY_PIPELINE_NAME = """
query PipelineSnapshotQueryByActivePipelineName($activePipelineSelector: PipelineSelector!) {
    pipelineSnapshotOrError(activePipelineSelector: $activePipelineSelector) {
        __typename
        ... on PipelineSnapshot {
            name
            pipelineSnapshotId
            description
            dagsterTypes { key }
            solids { name }
            modes { name }
            solidHandles { handleID }
            tags { key value }
            runTags { key value }
        }
        ... on PipelineSnapshotNotFoundError {
            snapshotId
        }
    }
}
"""


# makes snapshot tests much easier to debug
def pretty_dump(data) -> str:
    return json.dumps(data, indent=2, separators=(",", ": "))


def test_fetch_snapshot_or_error_by_snapshot_id_success(
    graphql_context: WorkspaceRequestContext, snapshot
):
    instance = graphql_context.instance
    result = noop_job.execute_in_process(instance=instance)
    assert result.success
    run = instance.get_run_by_id(result.run_id)
    assert run and run.job_snapshot_id

    result = execute_dagster_graphql(
        graphql_context,
        SNAPSHOT_OR_ERROR_QUERY_BY_SNAPSHOT_ID,
        {"snapshotId": run.job_snapshot_id},
    )

    assert not result.errors
    assert result.data
    assert result.data["pipelineSnapshotOrError"]["__typename"] == "PipelineSnapshot"

    snapshot.assert_match(pretty_dump(result.data))


def test_fetch_snapshot_or_error_by_snapshot_id_snapshot_not_found(
    graphql_context: WorkspaceRequestContext, snapshot
):
    result = execute_dagster_graphql(
        graphql_context,
        SNAPSHOT_OR_ERROR_QUERY_BY_SNAPSHOT_ID,
        {"snapshotId": "notthere"},
    )

    assert not result.errors
    assert result.data
    assert result.data["pipelineSnapshotOrError"]["__typename"] == "PipelineSnapshotNotFoundError"
    assert result.data["pipelineSnapshotOrError"]["snapshotId"] == "notthere"
    snapshot.assert_match(pretty_dump(result.data))


def test_fetch_snapshot_or_error_by_active_pipeline_name_success(
    graphql_context: WorkspaceRequestContext, snapshot
):
    result = execute_dagster_graphql(
        graphql_context,
        SNAPSHOT_OR_ERROR_QUERY_BY_PIPELINE_NAME,
        {
            "activePipelineSelector": {
                "pipelineName": "tagged_job",
                "repositoryName": main_repo_name(),
                "repositoryLocationName": main_repo_location_name(),
            }
        },
    )

    assert not result.errors
    assert result.data
    assert result.data["pipelineSnapshotOrError"]["__typename"] == "PipelineSnapshot"
    assert result.data["pipelineSnapshotOrError"]["name"] == "tagged_job"
    assert result.data["pipelineSnapshotOrError"]["tags"] == [{"key": "foo", "value": "bar"}]
    assert result.data["pipelineSnapshotOrError"]["runTags"] == [{"key": "baz", "value": "quux"}]

    snapshot.assert_match(pretty_dump(result.data))


def test_fetch_snapshot_or_error_by_active_pipeline_name_not_found(
    graphql_context: WorkspaceRequestContext, snapshot
):
    result = execute_dagster_graphql(
        graphql_context,
        SNAPSHOT_OR_ERROR_QUERY_BY_PIPELINE_NAME,
        {
            "activePipelineSelector": {
                "pipelineName": "jkdjfkdj",
                "repositoryName": main_repo_name(),
                "repositoryLocationName": main_repo_location_name(),
            }
        },
    )

    assert not result.errors
    assert result.data
    assert result.data["pipelineSnapshotOrError"]["__typename"] == "PipelineNotFoundError"

    snapshot.assert_match(pretty_dump(result.data))


def test_temporary_error_or_deletion_after_instance_check():
    instance = mock.MagicMock()

    instance.has_historical_job.return_value = True
    instance.get_historical_job.return_value = None

    with pytest.raises(UserFacingGraphQLError):
        _get_job_snapshot_from_instance(instance, "kjdkfjd")


def test_fetch_snapshot_or_error_by_snap_and_selector(
    graphql_context: WorkspaceRequestContext, snapshot
):
    # falls back to selector on snap id miss
    result = execute_dagster_graphql(
        graphql_context,
        SNAPSHOT_OR_ERROR_QUERY_BY_PIPELINE_NAME,
        {
            "snapshotId": "notthere",
            "activePipelineSelector": {
                "pipelineName": "csv_hello_world",
                "repositoryName": main_repo_name(),
                "repositoryLocationName": main_repo_location_name(),
            },
        },
    )

    assert not result.errors
    assert result.data
    assert result.data["pipelineSnapshotOrError"]["__typename"] == "PipelineSnapshot"
    assert result.data["pipelineSnapshotOrError"]["name"] == "csv_hello_world"

    instance = graphql_context.instance
    result = noop_job.execute_in_process(instance=instance)
    assert result.success
    run = instance.get_run_by_id(result.run_id)
    assert run and run.job_snapshot_id

    # valid snap id but bad selector works (snap preferred)
    result = execute_dagster_graphql(
        graphql_context,
        SNAPSHOT_OR_ERROR_QUERY_BY_SNAPSHOT_ID,
        {
            "snapshotId": run.job_snapshot_id,
            "activePipelineSelector": {
                "pipelineName": "does_not_exist",
                "repositoryName": main_repo_name(),
                "repositoryLocationName": main_repo_location_name(),
            },
        },
    )

    assert not result.errors
    assert result.data
    assert result.data["pipelineSnapshotOrError"]["__typename"] == "PipelineSnapshot"
