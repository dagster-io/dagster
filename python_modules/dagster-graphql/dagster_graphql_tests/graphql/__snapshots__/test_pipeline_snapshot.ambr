# serializer version: 1
# name: test_fetch_snapshot_or_error_by_active_pipeline_name_not_found[0]
  '''
  {
    "pipelineSnapshotOrError": {
      "__typename": "PipelineNotFoundError"
    }
  }
  '''
# ---
# name: test_fetch_snapshot_or_error_by_active_pipeline_name_success[0]
  '''
  {
    "pipelineSnapshotOrError": {
      "__typename": "PipelineSnapshot",
      "dagsterTypes": [
        {
          "key": "Any"
        },
        {
          "key": "Bool"
        },
        {
          "key": "Float"
        },
        {
          "key": "Int"
        },
        {
          "key": "Nothing"
        },
        {
          "key": "String"
        }
      ],
      "description": null,
      "modes": [
        {
          "name": "default"
        }
      ],
      "name": "tagged_job",
      "pipelineSnapshotId": "c3320c44d5fa6c508dfda564e67bc67747b9891c",
      "runTags": [
        {
          "key": "baz",
          "value": "quux"
        }
      ],
      "solidHandles": [
        {
          "handleID": "simple_op"
        }
      ],
      "solids": [
        {
          "name": "simple_op"
        }
      ],
      "tags": [
        {
          "key": "foo",
          "value": "bar"
        }
      ]
    }
  }
  '''
# ---
# name: test_fetch_snapshot_or_error_by_snapshot_id_snapshot_not_found[0]
  '''
  {
    "pipelineSnapshotOrError": {
      "__typename": "PipelineSnapshotNotFoundError",
      "snapshotId": "notthere"
    }
  }
  '''
# ---
# name: test_fetch_snapshot_or_error_by_snapshot_id_success[0]
  '''
  {
    "pipelineSnapshotOrError": {
      "__typename": "PipelineSnapshot",
      "dagsterTypes": [
        {
          "key": "Any"
        },
        {
          "key": "Bool"
        },
        {
          "key": "Float"
        },
        {
          "key": "Int"
        },
        {
          "key": "Nothing"
        },
        {
          "key": "String"
        }
      ],
      "description": null,
      "modes": [
        {
          "name": "default"
        }
      ],
      "name": "noop_job",
      "pipelineSnapshotId": "dbc0d71fefa48d56ff94e517235980f1bbf0d8e8",
      "runTags": [],
      "solidHandles": [
        {
          "handleID": "noop_op"
        }
      ],
      "solids": [
        {
          "name": "noop_op"
        }
      ],
      "tags": []
    }
  }
  '''
# ---
